Simple_Game.thy

```diff
@@ -9,64 +9,120 @@ text \<open>Simple games are games where player0 wins all infinite plays.\<close
+lemma strategy1_plays_subset:
+  assumes \<open>play \<in> plays_for_1strategy f initial\<close>
+  shows \<open>play \<in> plays initial\<close>
+  using assms by (induct rule: plays_for_1strategy.induct, auto simp add: plays.intros)
 
 lemma no_empty_plays:
-  assumes \<open>[] \<in> plays\<close>
+  assumes \<open>[] \<in> plays initial\<close>
   shows \<open>False\<close> 
   using assms plays.cases by blast
 
 text \<open>player1 wins a play if the play has reached a deadlock where it's player0's turn\<close>
 
-definition player1_wins :: \<open>'s play \<Rightarrow> bool\<close> where
-  \<open>player1_wins play \<equiv> player0_position (hd play) \<and> (\<nexists> p' . (hd play) \<longmapsto>\<heartsuit> p')\<close>
-
-definition player0_winning_strategy :: \<open>'s strategy \<Rightarrow> bool\<close> where
-  \<open>player0_winning_strategy f \<equiv> (\<forall> play \<in> plays_for_strategy f . \<not> player1_wins play)\<close>
+definition player1_wins_immediately :: \<open>'s play \<Rightarrow> bool\<close> where
+  \<open>player1_wins_immediately play \<equiv> player0_position (hd play) \<and> (\<nexists> p' . (hd play) \<longmapsto>\<heartsuit> p')\<close>
+
+definition player0_winning_strategy :: \<open>'s strategy \<Rightarrow> 's \<Rightarrow> bool\<close> where
+  \<open>player0_winning_strategy f initial \<equiv> (\<forall> play \<in> plays_for_0strategy f initial . \<not> player1_wins_immediately play)\<close>
+
+definition player0_wins :: \<open>'s \<Rightarrow> bool\<close> where
+  \<open>player0_wins s \<equiv> (\<exists> f . player0_winning_strategy f s \<and> sound_0strategy f s)\<close>
+
+lemma stuck_player0_win:
+  assumes \<open>player1_position initial\<close> \<open>(\<nexists> p' . initial \<longmapsto>\<heartsuit> p')\<close>
+  shows \<open>player0_wins initial\<close>
+proof -
+  have \<open>\<And>pl. pl \<in> plays initial \<Longrightarrow> pl = [initial]\<close>
+  proof -
+    fix pl
+    assume \<open>pl \<in> plays initial\<close>
+    thus \<open>pl = [initial]\<close> using assms(2) by (induct, auto)
+  qed
+  thus ?thesis using assms(1)
+    by (metis list.sel(1) player0_winning_strategy_def player0_wins_def player1_wins_immediately_def
+        sound_0strategy_def strategy0_plays_subset)
+qed
+
+
+definition player0_wins_immediately :: \<open>'s play \<Rightarrow> bool\<close> where
+  \<open>player0_wins_immediately play \<equiv> player1_position (hd play) \<and> (\<nexists> p' . (hd play) \<longmapsto>\<heartsuit> p')\<close>
+
+\<comment> \<open>in comparison to the defender, the attacker also has to bound the depth of plays.\<close>
+definition player1_winning_strategy :: \<open>'s strategy \<Rightarrow> 's \<Rightarrow> bool\<close> where
+  \<open>player1_winning_strategy f initial \<equiv>
+    \<exists>maxn. (\<forall> play \<in> plays_for_1strategy f initial . \<not> player0_wins_immediately play \<and> length play \<le> maxn)\<close>
```

Comment:
@luisamontanari It just occurred to me that one actually cannot always set a maximal natural number for each attacker-won position.

Consider a situation where the game has a defender position with countably-infinite branching and with the attacker definitely winning the *ith* branch after *i* steps. Clearly, the attacker wins at the branching position. Still, before the branching, you can't give a number of steps to expect.

The problem also carries over to the `HML_Strategies_Game` theory. But there, it should be easier to fix by switching from a decreasing measure of naturals to ordinals. (?)

Apparently, one must indeed explain the well-foundedness of attacker strategies more explicitly.